core/src/main/java/org/apache/iceberg/hadoop/HadoopStreams.java

```diff
@@ -185,8 +185,20 @@ public void flush() throws IOException {
 
     @Override
     public void close() throws IOException {
-      stream.close();
-      this.closed = true;
+      try {
+        stream.close();
+        // {@link org.apache.hadoop.fs.s3a.S3ABlockOutputStream#close()} calls {@link
+        // org.apache.hadoop.fs.s3a.S3ABlockOutputStream#putObject()}
+        // which doesn't throw an exception when interrupted. Need to handle interrupts here to make
+        // sure file is uploaded on close.
+        if (Thread.interrupted()
+            && "S3ABlockOutputStream"
+                .equals(stream.getWrappedStream().getClass().getSimpleName())) {
+          throw new IOException("object upload failed as it was interrupted during close");
```

Comment:
nit: technically, the upload was interrupted (not during close).
```suggestion
          throw new IOException("Failed to close stream since object upload failed.");
```
or something similar? Or maybe this exception should really be thrown by the write APIs?

Reply:
technically it was interrupted during close as:

`stream.close -> ... internally calls putObject() .. which uploads to s3`

Reply:
How about just "object upload interrupted"

Reply:
Done. Thanks